src/ooni/utils.hpp

```diff
@@ -5,34 +5,86 @@
 #ifndef SRC_OONI_UTILS_HPP
 #define SRC_OONI_UTILS_HPP
 
+#include "src/ext/json/src/json.hpp"
+#include <GeoIP.h>
 #include <measurement_kit/common.hpp>
 #include <measurement_kit/http.hpp>
 #include <regex>
 #include <string>
 
+using Json = nlohmann::json;
+
 namespace mk {
 namespace ooni {
 
 template <decltype(mk::http::get) httpget = mk::http::get>
 void ip_lookup(Callback<std::string> callback) {
     httpget("http://geoip.ubuntu.com/lookup",
-              [=](Error err, http::Response response) {
-                  if (err) {
-                      callback(err, "");
-                      return;
-                  }
-                  if (response.status_code != 200) {
-                      callback(GenericError(), "");
-                      return;
-                  }
-                  std::smatch m;
-                  std::regex regex("<Ip>(.*)</Ip>");
-                  if (std::regex_search(response.body, m, regex) == false) {
-                      callback(GenericError(), "");
-                      return;
-                  }
-                  callback(NoError(), m[1]);
-              }, {}, "", {}, Logger::global(), Poller::global());
+            [=](Error err, http::Response response) {
+                if (err) {
+                    callback(err, "");
+                    return;
+                }
+                if (response.status_code != 200) {
+                    callback(GenericError(), "");
+                    return;
+                }
+                std::smatch m;
+                std::regex regex("<Ip>(.*)</Ip>");
+                if (std::regex_search(response.body, m, regex) == false) {
+                    callback(GenericError(), "");
+                    return;
+                }
+                callback(NoError(), m[1]);
+            },
+            {}, "", {}, Logger::global(), Poller::global());
+}
+
+// this is needed to return a const char instead of a char for asn resolver
+inline GEOIP_API const char *
+MK_GeoIP_name_by_name_gl(GeoIP *gi, const char *host, GeoIPLookup *gl) {
+    return (const char *)GeoIP_name_by_name_gl(gi, host, gl);
+}
+
+template <decltype(GeoIP_country_code3_by_name_gl) resolver>
+std::string geoip_query(std::string ip, std::string path) {
+    GeoIP *gi;
+    GeoIPLookup gl;
+    gi = GeoIP_open(path.c_str(), GEOIP_MEMORY_CACHE);
+    if (gi == nullptr) {
+        return "";
+    }
+    const char *result;
+    result = resolver(gi, ip.c_str(), &gl);
+    if (result == NULL) {
+        return "";
+    }
+    GeoIP_delete(gi);
+    return std::string(result);
+}
+
+inline ErrorOr<Json> geoip(std::string ip, std::string path_country,
```

Comment:
If there's no need to templatize this function, I'd rather not inline it and have it in `ooni/utils.cpp`.